### PR TITLE
8336031: Create implementation of NSAccessibilityStaticText protocol

### PR DESCRIPTION
**Pull request by azuev-java**
Initial implementation of the protocol.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8336031](https://bugs.openjdk.org/browse/JDK-8336031): Create implementation of NSAccessibilityStaticText protocol (**Enhancement** - P4)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1507/head:pull/1507` \
`$ git checkout pull/1507`

Update a local copy of the PR: \
`$ git checkout pull/1507` \
`$ git pull https://git.openjdk.org/jfx.git pull/1507/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1507`

View PR using the GUI difftool: \
`$ git pr show -t 1507`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1507.diff">https://git.openjdk.org/jfx/pull/1507.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1507#issuecomment-2229457622)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Reviewers: @arapte @andy-goryachev-oracle 

/reviewers 2

File: modules/javafx.graphics/src/main/native-glass/mac/a11y/AccessibleBase.m [hunk 1]
```diff
@@ -50,6 +50,7 @@ + (void) initializeRolesMap {
     [rolesMap setObject:@"JFXRadiobuttonAccessibility" forKey:@"PAGE_ITEM"];
     [rolesMap setObject:@"JFXCheckboxAccessibility" forKey:@"CHECK_BOX"];
     [rolesMap setObject:@"JFXCheckboxAccessibility" forKey:@"TOGGLE_BUTTON"];
+    [rolesMap setObject:@"JFXStaticTextAccessibility" forKey:@"TEXT"];
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/native-glass/mac/a11y/AccessibleBase.m:53]: are these keys mapped to AccessibleRole.java enum?  if so,
1. should we add a comment stating that (and possibly add a comment in AccessibleRole as well?
2. are we going to keep adding roles to this block as we go along?

Comment (azuev-java) [reply modules/javafx.graphics/src/main/native-glass/mac/a11y/AccessibleBase.m:53]: 1. Yes, these are linked to the JFX accessible roles defined in that enum. I can add comment here - not sure about the AccessibleRole, it is a definition that might be used for more than one platform so linking platform specific implementation from there sound incorrect;
2. Yes, as we are converting more and more functionality to the new API we are going to add roles implementations to this block until we have no roles not covered with the new API.

Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/native-glass/mac/a11y/AccessibleBase.m:55]: thank you for clarifying.  fully agree with #1, referencing from platform specific to the common should be sufficient.

**Comment by arapte** [comment]
I observed a crash in Ensemble.
1. Launch VoiceOver
2. Launch Ensemble
3. Click on the Stacked Area Chart sample, (thumbnail shown in the beginning of Ensemble app home page)
-> Ensemble crashes with following error

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff80138665d, pid=69749, tid=259
#
# JRE version: OpenJDK Runtime Environment (21.0.1+12) (build 21.0.1+12-29)
# Java VM: OpenJDK 64-Bit Server VM (21.0.1+12-29, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libobjc.A.dylib+0x665d]  objc_msgSend+0x1d
```
Callstack from hs_err*.log

```
Current thread (0x00007fbad9a12a00):  JavaThread "JavaFX Application Thread"        [_thread_in_native, id=259, stack(0x00007ff7bc8d3000,0x00007ff7bd0d3000) (8192K)]

Stack: [0x00007ff7bc8d3000,0x00007ff7bd0d3000],  sp=0x00007ff7bd0cbe88,  free space=8163k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libobjc.A.dylib+0x665d]  objc_msgSend+0x1d
C  [libglass.dylib+0x8b87]  Java_com_sun_glass_ui_mac_MacAccessible_isEqualToString+0x17
J 3994  com.sun.glass.ui.mac.MacAccessible.isEqualToString(JJ)Z javafx.graphics@24-internal (0 bytes) @ 0x00000001158d8e6f [0x00000001158d8da0+0x00000000000000cf]
J 4043 c1 com.sun.glass.ui.mac.MacAccessible$MacAttribute.getAttribute(J)Lcom/sun/glass/ui/mac/MacAccessible$MacAttribute; javafx.graphics@24-internal (63 bytes) @ 0x000000010e3bb7f4 [0x000000010e3bb500+0x00000000000002f4]
j  com.sun.glass.ui.mac.MacAccessible.accessibilityAttributeValueForParameter(JJ)Lcom/sun/glass/ui/mac/MacVariant;+4 javafx.graphics@24-internal
v  ~StubRoutines::call_stub 0x0000000115199cc6
V  [libjvm.dylib+0x5abd24]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x2b4
V  [libjvm.dylib+0x615ecf]  jni_invoke_nonstatic(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, JavaThread*)+0x34f
V  [libjvm.dylib+0x618223]  jni_CallLongMethod+0x193
C  [libglass.dylib+0x2fa34]  -[JFXStaticTextAccessibility accessibilityAttributedStringForRange:]+0x114
C  [AppKit+0x7a3d01]  -[NSAccessibilityAttributeAccessorInfo getParameterizedAttributeValue:forObject:withParameter:]+0x41
C  [AppKit+0x7a5b4c]  ___NSAccessibilityEntryPointValueForAttributeWithParameter_block_invoke.888+0x1bf
C  [AppKit+0x7a08bb]  NSAccessibilityPerformEntryPointObject+0x10
C  [AppKit+0x7a24cb]  NSAccessibilityEntryPointValueForAttributeWithParameter+0xac
C  [AppKit+0x4d7a9b]  CopyParameterizedAttributeValue+0x1ed
C  [HIServices+0x27e58]  _AXXMIGCopyParameterizedAttributeValue+0x1da
C  [HIServices+0x4bcfd]  _XCopyParameterizedAttributeValue+0x250
C  [HIServices+0x8363]  mshMIGPerform+0xbb
C  [CoreFoundation+0x7c5b7]  __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__+0x29
C  [CoreFoundation+0x7c4ee]  __CFRunLoopDoSource1+0x216
C  [CoreFoundation+0x7b166]  __CFRunLoopRun+0xa8c
C  [CoreFoundation+0x7a112]  CFRunLoopRunSpecific+0x22d
C  [HIToolbox+0x2ea09]  RunCurrentEventLoopInMode+0x124
C  [HIToolbox+0x2e816]  ReceiveNextEventCommon+0x299
C  [HIToolbox+0x2e561]  _BlockUntilNextEventMatchingListInModeWithFilter+0x42
C  [AppKit+0x3ec61]  _DPSNextEvent+0x370
C  [AppKit+0x952dc0]  -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:]+0x4f9
C  [AppKit+0x30075]  -[NSApplication run]+0x25b
C  [libglass.dylib+0xae72]  -[GlassApplication runLoop:]+0x7f2
C  [Foundation+0x7d365]  __NSThreadPerformPerform+0xb2
C  [CoreFoundation+0x7c087]  __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__+0x11
C  [CoreFoundation+0x7c029]  __CFRunLoopDoSource0+0x9d
C  [CoreFoundation+0x7bdf4]  __CFRunLoopDoSources0+0xd7
C  [CoreFoundation+0x7aa71]  __CFRunLoopRun+0x397
C  [CoreFoundation+0x7a112]  CFRunLoopRunSpecific+0x22d
C  [libjli.dylib+0x6ee2]  CreateExecutionEnvironment+0x182
C  [libjli.dylib+0x28bd]  JLI_Launch+0x54d
C  [java+0x3c0f]  main+0x18f
C  [dyld+0x6366]  start+0x796
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 3994  com.sun.glass.ui.mac.MacAccessible.isEqualToString(JJ)Z javafx.graphics@24-internal (0 bytes) @ 0x00000001158d8dfc [0x00000001158d8da0+0x000000000000005c]
J 4043 c1 com.sun.glass.ui.mac.MacAccessible$MacAttribute.getAttribute(J)Lcom/sun/glass/ui/mac/MacAccessible$MacAttribute; javafx.graphics@24-internal (63 bytes) @ 0x000000010e3bb7f4 [0x000000010e3bb500+0x00000000000002f4]
j  com.sun.glass.ui.mac.MacAccessible.accessibilityAttributeValueForParameter(JJ)Lcom/sun/glass/ui/mac/MacVariant;+4 javafx.graphics@24-internal
v  ~StubRoutines::call_stub 0x0000000115199cc6
```

**Comment by azuev-java** [comment]
Fixed the crash by eliminating request to unsupported type of metadata.

**Review by arapte — APPROVED** [review]

**Comment by andy-goryachev-oracle** [comment]
I don't see crashes, but I noticed a few things when testing with the Monkey Tester:

1. on the AreaChart page, I was unable to see the a11y focus rectangle around the elements of the chart, even though I can hear what sounded like announcements for various elements on the chart.  This behavior differed from e.g. BarChart where I could navigate with a11y focus using the TAB key.  What's more interesting, this behavior appears to be the same regardless of the 'focus traversable' enabled property
2. On the BarChart (and on other chart pages) I could not place a11y focus on any of the chart  elements using the mouse; but the TAB key worked.  Why is that?
3. HtmlEditor: the focus rectangle is fixed at the small area in the top left corner, regardless of the text entered (see the screenshot)

Screenshot 2024-08-06 at 07 44 42

4. In the WebView page, enter the https://github.com address and press enter.  I see the following output in stderr (which is absent with VoiceOver off):
```
Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread" state:SCHEDULED
state:RUNNING
```

**Comment by kevinrushforth** [comment]
@andy-goryachev-oracle  Can you run the same tests with mainline (i.e., without the fix from this PR)? Those problems that only occur with this patch should be addressed as part of this PR. Problems that occur both with and without this patch should be tracked separately.

**Comment by andy-goryachev-oracle** [comment]
> Can you run the same tests with mainline

Items 1-3 can be seen with mainline, so these might be (new) issues.

Item 4 seems to have been introduced by this PR - don't see the issue with the master branch.  Also, notice that we don't get a stack trace but a bunch of concatenated messages.

**Comment by azuev-java** [comment]
> Item 4 seems to have been introduced by this PR - don't see the issue with the master branch. Also, notice that we don't get a stack trace but a bunch of concatenated messages.

I can not reproduce this on my build - probably because i use Ensemble to test WebView, i will try again with MonkeyTester - but with Ensemble even without my fix i see a bunch of exceptions just by scrolling with my trackpad (without VoiceOver so not caused by the a11y code) that look like this:

SEVERE: RenderJob error
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "com.sun.prism.RTTexture.contentsUseful()" because "this.txt" is null
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at javafx.web@24-internal/com.sun.javafx.webkit.prism.PrismInvoker.runOnRenderThread(PrismInvoker.java:93)
...

so i just wonder if there is something with github site which WebView can not load or parse which leads to the displayed structure being not correct hence errors when a11y code fetches properties of the text elements.

**Review by andy-goryachev-oracle — APPROVED** [review]
looks good, tested with the monkey tester on macOS 14.6.1 M1
I don't see the exception or stderr anymore.

This output is seen (as expected):
```
WARNING: A native method in com.sun.webkit.WebPage has been bound
WARNING: com.sun.webkit.WebPage::twkInitWebCore is declared in module javafx.web (file:/Users/angorya/Projects/jfx-1/jfx/rt/modules/javafx.web/bin/)
WARNING: Use --enable-native-access=javafx.web to avoid a warning for native methods declared in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled
```

**Pull request closed without merging**